kernel/reduction.ml

```diff
@@ -1061,6 +1061,38 @@ let dest_lam_assum env =
   in
   lamec_rec env Context.Rel.empty
 
+let dest_prod_n_assum env n =
+  let rec prodec_rec env n l ty =
+    if Int.equal n 0 then l,ty
+    else
+    let rty = whd_allnolet env ty in
+    match kind rty with
+    | Prod (x,t,c)  ->
+        let d = LocalAssum (x,t) in
+        prodec_rec (push_rel d env) (n-1) (Context.Rel.add d l) c
+    | LetIn (x,b,t,c) ->
+        let d = LocalDef (x,b,t) in
+        prodec_rec (push_rel d env) (n-1) (Context.Rel.add d l) c
```

Comment:
I have a doubt here, shouldn't it be `n` rather than `n-1` as we're only counting assumptions?
Which one is called from inductive.ml?

Reply:
I copy-pasted `decompose_prod_n_assum` which has the `n-1`. But you're right, this is not consistent with the `assum` suffix and not consistent either with `decompose_lambda_n_assum` which have the `n` in the let case.

I then removed `dest_prod_n_assum` from the PR and used `n` in `dest_lam_n_assum` (I copied the latter from `dest_prod_n_assum` without realizing the problem, sorry).

I'm going to make a proposal below for clarifying the `n_assum` business in another PR.

Reply:
Typical let-binding confusion, I would not blame you I believe only verified code can get it right everywhere :)